refactor: Changes the DatabaseSelector and TableSelector to use the new Select component

michael-s-molina opened the pull request:
### SUMMARY
Follow-up of https://github.com/apache/superset/pull/16334 which introduced https://github.com/apache/superset/issues/16475 and was reverted in https://github.com/apache/superset/pull/16478.

You can find the descriptions and videos of the changes in the original PR https://github.com/apache/superset/pull/16334.

Fixes https://github.com/apache/superset/issues/16475

Thanks, @etr2460 for providing the revert PR.

@ktmud @etr2460 @villebro
I'll add comments indicating the relevant changes from the original PR.

Is also worth mentioning this fix by @AAfghahi https://github.com/apache/superset/pull/16472

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Check the original PR.

### TESTING INSTRUCTIONS
Check the original PR.

### ADDITIONAL INFORMATION
- [x] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
refactor: Changes the DatabaseSelector and TableSelector to use the new Select component

etr2460 reviewed (commented):
2 questions about testing/verification

michael-s-molina commented on superset-frontend/src/components/DatabaseSelector/DatabaseSelector.test.tsx:
```diff
@@ -153,50 +153,62 @@ test('Refresh should work', async () => {
+  expect(select).toBeInTheDocument();
+  userEvent.click(select);
+  expect(
+    await screen.findByRole('option', { name: 'postgresql: test' }),
```
Great suggestion. I created two new tests to catch the previous regression:
- Sends the correct db when changing the database
- Sends the correct schema when changing the schema

michael-s-molina commented on superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx:
```diff
@@ -227,10 +227,7 @@ class DatasourceControl extends React.PureComponent {
           {extra?.warning_markdown && (
-            <WarningIconWithTooltip
-              warningMarkdown={extra.warning_markdown}
-              size={30}
```
Of course. I'm attaching a screenshot with both `healthCheckMessage` and `warning_markdown` enabled for comparison.

<img width="295" alt="Screen Shot 2021-09-01 at 12 01 19 PM" src="https://user-images.githubusercontent.com/70410625/131695472-bdc8055d-e45f-41d4-8445-622b36820fe4.png">

michael-s-molina commented:
> 1. For db types, can we use the same tag style instead of just text?

Done

> 2. Loading state should show "Loading..." instead of "No data"

Done in https://github.com/apache/superset/pull/16531

> 3. Previously the Select will automatically trigger the initial load of the tables list if there is a selected schema and the schema list if there is a selected database (just select a table and refresh the page to test this), now you have to focus on the tables select to trigger the loading. I think we should keep this preload behavior because it can be quite slow to load these APIs.

The idea here is to avoid unnecessary queries to the server. This component is used in Explore, SQL Editor, and the datasets list. Many times the user does not modify any of the selects but they are rendered. Some examples:
- When the user enters the datasets modal from the dataset list to edit anything from the other tabs
- When changing or just viewing a query in SQL Editor

I do understand your concern though. Here we have a limitation because the tables endpoint does not currently support pagination. In talks with @villebro to add pagination support we ended up with the following:

> That tables API will be horrible to paginate, as SqlAlchemy doesn’t natively support pagination of retrieval of table names.
I can open up a discussion on the sqla repo to see if they're open to adding the optional feature, but it will probably take a long time to get it in.

So one of the reasons for the slow behavior is the lack of pagination support. For cases where the queries are heavy even with pagination, we created a prop called `fetchOnlyOnSearch` that only triggers a filtered query. I think Airbnb should enable this prop for this case given that you have a schema with 30k tables.

> 4. Scrolling tables list to the end will trigger an unnecessary new query----probably trying to fetch the next page?

This was a result of the lack of pagination support. The component thinks that we have more results available given the page size and total results. To overcome this, while we don't add pagination support to this endpoint, I added the following:

```
// TODO: Would be nice to add pagination in a follow-up. Needs endpoint changes.
pageSize={Number.MAX_SAFE_INTEGER}
```

This is just disabling the pagination in practice. We don't want to provide a prop to disable the pagination because we want to make sure all endpoints support this feature. This is especially important if we think about devices with slow connections that are using Superset endpoints.

> It also freezes the whole page on our one of our schemas with 30k tables.

I tested here with 50k values and the Select was able to render smoothly. Virtualization is native in Ant Design Select. In their page, we have an [example](https://ant.design/components/select/#API) with 100k items.  So I don't know if it froze because of query performance or another reason. Maybe we can test again with `fetchOnlyOnSearch` enabled.

> 5. Switching schema does not clear the selected table--even though the table does not exist in the new schema.

Fixed.

@ktmud @etr2460 @villebro Thank you so much for the tests, reviews and help improving this feature.

ktmud commented:
Table names and schema names are cached, so I don't think unnecessary db metadata queries are a huge concern.

michael-s-molina commented:
@ktmud Let me provide you a little more context to help with the decision. When we started the select migration project, @geido compiled a great spreadsheet with all the different types of selects that we had and all the props that were available for each select type. We had a lot of props, more than 50! What we noticed was that this lead to multiple behaviors throughout the application and increased the complexity of components because of the different combinations. Then we paired with @villebro to design the component API and decided that we would try to diminish the number of props by establishing default behaviors and making the user experience more similar. Another objective was to decrease the complexity of the component.

This context is important because we can resolve these two questions by adding new props to the component like `lazyLoading` (default true) and `paginated` (default true). In fact, these props appeared during our discussions but we assumed that we would always want to avoid unnecessary queries and make sure that data is always fetched in blocks. That's why they are default behaviors of the component and not props.

> Table names and schema names are cached, so I don't think unnecessary db metadata queries are a huge concern.

This can be resolved with the inclusion of the `lazyLoading` prop, but I don't think that even if the lazy loading is disabled we will resolve the problem. The query can still be running when the user opens the select, even if it started during the rendering phase. So I think using `fetchOnlyOnSearch` would be more appropriate for this case.

> Re: pagination, shouldn't the Select itself has a built-in "async but no pagination" mode? Why is the pageSize hack necessary?

The hack here is needed because the default behavior assumes that we should always try to fetch data using pagination.  Adding a `paginated` prop can encourage the developer to not deal with adding pagination to the endpoints, which is pretty common right now. What the hack does is establishing a really big page size so that the content fits on one page while we don't fix the endpoint.

As these are subjective decisions, I still can add these props if you think they pay off. I just wanted to add the whole context to help with the decision.

ktmud commented:
> The query can still be running when the user opens the select, even if it started during the rendering phase. 

This is about improving the availability of UI elements for most cases. A lot of users land on SQL Lab page without immediately going to change the selected table, but when they do, it's better to have the list preloaded already.

TBH I don't think pagination and async loading itself should not be part of the default `Select` component. A lot of select controls do not need async loading at all (but still need the `isCreatable` and `isMultible` options).

The Async behavior should either be a HOC or another hook. For example:

```ts
const { options, onSearch } = useAsyncSelectOptions({
  fetchOptions: ...,
  preload: true
});

<Select options={options} onSearch={onSearch} .../>
```

With the complexity abstracted away from the Select component itself, adding more options to the fetch behavior should not be that big of a concern.

michael-s-molina commented:
> The Async behavior should either be an HOC or another hook.

We consider this approach when designing the component but we chose to make it an integral part of the Select because we have UI behaviors that are more easily implemented as a native feature than as a HOC. It's not only about how to retrieve the data but the interaction with the component also changes.

> A lot of select controls do not need async loading at all.

We just replaced all the selects of the application and the majority of them require async loading. In many cases where the previous select received an options array, it was the parent component that was doing the fetch and managing the pagination. In fact, absorbing this behavior and avoiding this duplicated logic was one of the reasons to create the component in the first place.

I think that in the interest of moving this PR forward, can we settle in creating the `paginated` and `lazyLoading` props for now and revisit this decision when all endpoints support pagination?

michael-s-molina commented:
> Not sure if this would help, but recently I've created [yet another Select component](https://github.com/ktmud/chakra-ui-downshift-select/blob/main/src/Select/MultiSelect.tsx) in another project using the amazing [Downshift.js](https://www.downshift-js.com/use-multiple-selection) library and [chakra-ui](https://chakra-ui.com/). It does not handle pagination yet, but the "letting hooks take over event handlers" approach that Downshift.js has taken really opened my eyes. Thought that would be relevant to the discussion here as well.
> 
> This Select component also uses [a custom render hook](https://github.com/ktmud/chakra-ui-downshift-select/blob/main/src/hooks/useRenderers.tsx) which has made it easy to break down a complex component into overridable smaller subcomponents.

Thanks for sharing this @ktmud. I'll definitely take a look for future improvements to the component. In the meantime, I removed the pagination and lazy loading, reverting the behavior to what it was before. We can continue with the review process.

michael-s-molina commented:
/testenv up

geido commented:
/testenv up

geido reviewed (approved):
Code LGTM! Thank you!

michael-s-molina commented:
/testenv up

etr2460 reviewed (approved):
lgtm!
